pkg/controller/endpointslice/utils.go

```diff
@@ -0,0 +1,252 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package endpointslice
+
+import (
+	"crypto/md5"
+	"encoding/hex"
+	"fmt"
+	"reflect"
+	"sort"
+	"time"
+
+	"github.com/davecgh/go-spew/spew"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	discovery "k8s.io/api/discovery/v1alpha1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/klog"
+	podutil "k8s.io/kubernetes/pkg/api/v1/pod"
+	api "k8s.io/kubernetes/pkg/apis/core"
+)
+
+// podEndpointChanged returns true if the results of podToEndpoint are different
+// for the pods passed to this function.
+func podEndpointChanged(pod1, pod2 *corev1.Pod) bool {
+	endpoint1 := podToEndpoint(pod1, &corev1.Node{})
+	endpoint2 := podToEndpoint(pod2, &corev1.Node{})
+
+	endpoint1.TargetRef.ResourceVersion = ""
+	endpoint2.TargetRef.ResourceVersion = ""
+
+	return !reflect.DeepEqual(endpoint1, endpoint2)
+}
+
+// podToEndpoint returns an Endpoint object generated from a Pod and Node.
+func podToEndpoint(pod *corev1.Pod, node *corev1.Node) discovery.Endpoint {
+	topology := map[string]string{}
+
+	if pod.Spec.NodeName != "" {
+		topology["kubernetes.io/node"] = pod.Spec.NodeName
+	}
+
+	if pod.Spec.Hostname != "" {
+		topology["kubernetes.io/pod-hostname"] = pod.Spec.Hostname
+	}
+
+	if node != nil {
+		topologyLabels := []string{
+			"failure-domain.beta.kubernetes.io/zone",
+			"failure-domain.beta.kubernetes.io/region",
+		}
+
+		for _, topologyLabel := range topologyLabels {
+			if node.Labels[topologyLabel] != "" {
+				topology[topologyLabel] = node.Labels[topologyLabel]
```

Comment:
Can we maintain a reverse lookup map of the pod to the associated service or service's topology keys like `map[metav1.UID]*corev1.Service` and `map[metav1.UID]map[string]string` in EndpointSlice Controller ? Then we we can get all the topology keys contained in the service associated with this pod. This makes it possible for user-defined topology keys other than use well-known topology keys only.
You can get topology keys from service once my service topology implementation PR https://github.com/kubernetes/kubernetes/pull/72046 been merged.

cc @khenidak @thockin @freehan

Reply:
Yes, let me think more, see if there is a better way.

Reply:
What about just cache a map of pod to a slice of associated services like`map[metav1.UID][]*v1.Service` ?
We will always call `GetPodServices` under the hook when pod add, update or delete right now:
``` go
services, err := serviceLister.GetPodServices(pod)
```
So we can cache this relationship at here, just a little memory cost I think (mostly pointers), and then we can traverse the `toplogyKeys` of all the services of the pod (automatically overwrite the duplicate toplogy key) when construct the `Topology` of `discovery.Endpoint`